addons/pos_loyalty/models/pos_order.py

```diff
@@ -144,6 +147,20 @@ def confirm_coupon_programs(self, coupon_data):
             'coupon_report': coupon_per_report,
         }
 
+    def _check_existing_loyalty_cards(self, coupon_data):
+        coupon_key_to_modify = []
+        for coupon_id, coupon_vals in coupon_data.items():
+            partner_coupons = self.env['loyalty.card'].search(
+                [('partner_id', '=', coupon_vals.get('partner_id', False)), ('program_type', '=', 'loyalty')])
+            existing_coupon_for_program = partner_coupons.filtered(lambda c: c.program_id.id == coupon_vals['program_id'])
+            if len(existing_coupon_for_program) > 1:
+                raise UserError(_('You cannot have multiple loyalty cards for the same program. Please delete the extra ones.'))
```

Comment:
The problem with this approach is that if there are multiple loyalty cards, none will be updated.

Instead of raising an error here, we continue with the logic and we update the first existing coupon. Then, in the UI, we should tell the cashier that there are multiple loyalty cards for the customer and that they need to merge it.